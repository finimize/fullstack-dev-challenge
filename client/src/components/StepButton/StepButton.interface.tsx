export interface StepButtonInterface {
    selected?: boolean
    label: number
}
