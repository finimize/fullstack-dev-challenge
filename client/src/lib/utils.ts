import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
