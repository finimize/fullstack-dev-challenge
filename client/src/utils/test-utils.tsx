import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
