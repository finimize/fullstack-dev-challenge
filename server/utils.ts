const compoundedInterest = (interestRate: number, timeInMonths: number) :number => {
    return Math.pow((interestRate + 1), timeInMonths);
};

const percentageToDecimal = (percentValue: number) :number => percentValue/100;

export const monthlyCompoundInterest = (monthlyDeposit: number, monthlyInterest: number, timeInMonths: number) :number =>  {
    const interestRate = percentageToDecimal(monthlyInterest);
    const interestOverMonths = compoundedInterest(interestRate, timeInMonths);
    const monthlyCompound = (monthlyDeposit *((interestOverMonths-1)/interestRate));
    return Math.round(monthlyCompound);
}

export const principleCompoundInterest = (initialDeposit: number,
                                          monthlyInterest: number,
                                          timeInMonths: number) : number => {
    const interestOverMonths = compoundedInterest(percentageToDecimal(monthlyInterest), timeInMonths);
    return Math.round(initialDeposit * interestOverMonths);
}

export const totalSavingsOverTime = (initialDeposit: number,
                                    monthlyDeposit: number,
                                    monthlyInterest: number,
                                    timeInMonths: number) :number => {
    const principle = principleCompoundInterest(initialDeposit, monthlyInterest, timeInMonths);
    const  monthly = monthlyCompoundInterest(monthlyDeposit, monthlyInterest, timeInMonths);
    return monthly+principle;
}

const toRange = (value: number) => {
    let array = [];
    for(let i =1;i<=value;i++){
        array.push(i);
    }
    return array;
}
export const toYearsFromNow = (years: number) :Array<string> => {
    let yearsArray = toRange(years);
    return yearsArray.map(number => {
        let d = new Date();
        let currentYear = d.getFullYear();
        let month = d.getMonth();
        let day = d.getDate();
        let c = new Date(currentYear + number, month, day);
        return c.getFullYear().toString();
    })
}
export const yearlySavingsOverTime = (initialDeposit: number,
                                        monthlyDeposit: number,
                                         monthlyInterest: number,
                                         timeInYears: number) : Array<number> => {
        const arr = toRange(timeInYears);
        const principleCompound = principleCompoundInterest(initialDeposit,monthlyInterest,timeInYears*12);
        return arr.map(number => monthlyCompoundInterest(monthlyDeposit,monthlyInterest,number*12) + principleCompound);
}
