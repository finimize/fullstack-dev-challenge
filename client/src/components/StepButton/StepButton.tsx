import React, { FC } from 'react'
import { Button } from '@chakra-ui/react'
import { StepButtonInterface } from './StepButton.interface'

export const StepButton: FC<StepButtonInterface> = ({ selected, label, disabled, onClick }) => {
    const hoverStyles = !disabled
        ? {
              borderColor: 'blueHeader',
              transform: `scale(1.1)`,
          }
        : {}
    return (
        <Button
            data-testid={`step-button-${label}`}
            onClick={onClick}
            disabled={disabled || false}
            borderRadius='50%'
            height='40px'
            width='40px'
            border='4px'
            borderColor={selected ? 'blueHeader' : 'primary'}
            color='blueHeader'
            fontSize='lg'
            cursor='pointer'
            margin='10px'
            _hover={{ ...hoverStyles }}
        >
            {label}
        </Button>
    )
}
