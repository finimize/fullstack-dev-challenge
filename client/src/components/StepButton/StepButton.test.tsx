import React from 'react'
import { cleanup } from '@testing-library/react'
import { customRender } from '../../utils/test-utils'
import { StepButton } from './StepButton'

afterEach(cleanup)

describe('component/Savings', () => {
    it('renders step button component', () => {
        const { render } = customRender(<StepButton label={1} />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('renders step button component when selected', () => {
        const { render } = customRender(<StepButton label={1} selected />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })
})
