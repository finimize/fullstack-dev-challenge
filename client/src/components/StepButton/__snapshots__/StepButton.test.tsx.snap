// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Savings renders Calculator Mode component 1`] = `
<div>
   
  <button
    class="chakra-button css-mxlia0"
    type="button"
  >
    1
  </button>
   
   
</div>
`;

exports[`component/Savings renders step button component 1`] = `
<div>
   
  <button
    class="chakra-button css-mxlia0"
    type="button"
  >
    1
  </button>
   
   
</div>
`;

exports[`component/Savings renders step button component when selected 1`] = `
<div>
   
  <button
    class="chakra-button css-10dvod4"
    type="button"
  >
    1
  </button>
   
   
</div>
`;
